docs: Documentation and clean up of bevy_math.

KDecay opened the pull request:
This PR is part of the issue #3492.
Closes #335.
Closes #3484.

# Objective
- Add and update the bevy_math documentation to achieve a 100% documentation coverage.
- Add the #![warn(missing_docs)] lint to keep the documentation coverage for the future.
- Remove the face_toward.rs file containing the `FaceToward` trait, because it is unused.
- Move the geometry.rs file to `bevy_ui` and rename the `Rect` type to `UiRect` to remove having two types with the same name (see `bevy_sprite::Rect`).
- Remove `bevy_ui::Margins`.
- Add a `new(left: T, right: T, top: T, bottom: T)` function for `bevy_ui::UiRect`.
- Change `Size<T>` to `Size<Val>` removing the generic, because Val should be extendable enough.
- Change `UiRect<T>` to `UiRect<Val>` removing the generic, because Val should be extendable enough.
- Document the behaviour of different position settings to clarify how the current UI implementation will react.
- Check if `bevy_math` works with the `doc_markdown` lint.
- Check if `bevy_ui` works with the `doc_markdown` lint.

## Solution
- [x] Add and update the bevy_math documentation.
- [x] Add the #![warn(missing_docs)] lint.
- [x] Remove the face_toward.rs file
- [x] Move the geometry.rs file to `bevy_ui`
- [x] Remove `bevy_ui::Margins`.
- [x] Add a `new(left: T, right: T, top: T, bottom: T)` function for `bevy_ui::UiRect`.
- [x] Change `Size<T>` to `Size<Val>`.
- [x]  Change `UiRect<T>` to `UiRect<Val>`.
- [x] Document the behaviour of different position settings.
- [x] Run `cargo clippy -p bevy_math --no-deps --all-targets --all-features -- -D warnings -A clippy::type_complexity -W clippy::doc_markdown` without errors.
- [x] Run `cargo clippy -p bevy_ui --no-deps --all-targets --all-features -- -D warnings -A clippy::type_complexity -W clippy::doc_markdown` without errors.

alice-i-cecile commented on crates/bevy_math/src/face_toward.rs:
```diff
@@ -1,8 +1,51 @@
 use crate::{Mat4, Vec3};
 
-/// Generates a translation / rotation matrix that faces a given target
+/// Implements a function to generate a translation / rotation matrix that faces a given target.
```
```suggestion
/// Implements a function that generates a translation / rotation matrix that faces a given target.
```

I think this reads slightly more clearly.

KDecay replied:
Done.

alice-i-cecile commented on crates/bevy_math/src/geometry.rs:
```diff
@@ -25,17 +55,55 @@ impl<T: Default + Reflect + PartialEq> Default for Size<T> {
-/// A rect, as defined by its "side" locations
+/// A rectangle defined by its side locations.
+///
+/// It is commonly used to define positions, margins, paddings and borders.
```
I think we should explain that the values supplied define the distances from the nominal "center" of the rect to the edge.

If any of the distances are negative, the center will not be contained within the rect.

KDecay replied:
Implemented this and also added a description of the example to clarify how the `Rect` works.

alice-i-cecile commented on crates/bevy_math/src/geometry.rs:
```diff
@@ -25,17 +55,55 @@ impl<T: Default + Reflect + PartialEq> Default for Size<T> {
 }
 
 impl<T: Reflect + PartialEq> Rect<T> {
+    /// Creates a new [`Rect`] from a single value.
```
```suggestion
    /// Creates a new [`Rect`] where all sides are equidistant from the center.
```

KDecay replied:
Done.

alice-i-cecile commented on crates/bevy_math/src/geometry.rs:
````diff
@@ -25,17 +55,55 @@ impl<T: Default + Reflect + PartialEq> Default for Size<T> {
+    /// assert_eq!(rect.top, 1.0);
+    /// assert_eq!(rect.bottom, 1.0);
+    /// ```
     pub fn all(value: T) -> Self
````
```suggestion
    pub fn all(distance_from_center: T) -> Self
```

When doing docs PRs, I often sneak in small variable renames like this that clarify intent.

KDecay replied:
Good Point! Done.

alice-i-cecile reviewed (changes requested):
This is excellent, thank you! I left a few comments with my thoughts, let me know what you think.

mockersf commented on crates/bevy_math/src/face_toward.rs:
````diff
@@ -1,8 +1,51 @@
+///
+/// ```rust
+/// # use bevy_math::{FaceToward, Vec3, Vec4};
+/// # #[allow(dead_code)]
````
is the `#[allow(dead_code)]` still needed? it seems it's used below

KDecay replied:
It was throwing warnings when I was testing the example in a separate crate. Since all the bevy_math modules are likely to be deleted/moved this documentation will be gone as well. I'll mark this as resolved and get started with cleaning up the bevy_math crate.

mockersf commented on crates/bevy_math/src/lib.rs:
````diff
@@ -1,10 +1,29 @@
+//!
+//! The commonly used types are vectors, rectangles and sizes.
+//!
+//! # Example
+//!
+//! ```rust
+//! use bevy_math::prelude::*;
+//! ```
+
````
I don't think this code example is really useful, and we don't want to encourage using `bevy_math` standalone

KDecay replied:
Understandable. I wasn't sure if the bevy crates should be documented as standalone or just assume that everyone is using bevy as a whole (which probably is the more likely case). I had a version where I had the following code which didn't work, because bevy is not a dependency of bevy_math. Adding the ignore keyword would make it harder to keep this up to date if we change something about the path in the future.

```rust
use bevy::math::prelude::*;
```

alice-i-cecile replied:
Yeah, there's only a few crates that are worth using standalone: namely `bevy_ecs` and `bevy_reflect`. Everything else is a bit nonsensical outside of the Bevy ecosystem.

KDecay replied:
Makes sense. I'll remove these examples then as stated below.

mockersf commented on crates/bevy_math/src/lib.rs:
````diff
@@ -1,10 +1,29 @@
 
+/// The bevy_math prelude.
+///
+/// # Example
+///
+/// ```rust
+/// use bevy_math::prelude::*;
+/// ```
````
Same for this one

KDecay replied:
Explanation in the conversation above.

mockersf commented:
Is the `FaceToward` trait really needed? It seems it's only duplicate functionality from https://docs.rs/glam/latest/glam/f32/struct.Mat4.html#method.look_at_rh

Reading the doc on `Rect`, I find it very confusing. Looking through its usage in code:
* first thing I noticed, we actually have two different structs named `Rect`, the other one is defined in `bevy_sprite` and has a much more logical definition
https://github.com/bevyengine/bevy/blob/1d0d8a3397bd6fc2c14d42ffd0668d2443748912/crates/bevy_sprite/src/rect.rs#L8-L13
* the `Rect` struct in `bevy_math` is, as you noted, only used in `bevy_ui`, and it doesn't actually define a rectangle, only what's to do to each side of one (with of a border, width of a padding, ...)

I think the `Rect` from `bevy_math` should be moved to `bevy_ui` (and renamed), and it could probably lose the type parameter as I think it's only used with `Val`.

There is a pretty similar situation with the `Size` struct in `bevy_math`. It's only used as a generic `Vec2`, and that's only needed in `bevy_ui` as it's needed to have a `Vec2` over `Val`, which is not possible with the `Vec2` from `glam`. Most other places in code that need a size use directly a `Vec2`.

I think the `Size` struct should move to `bevy_ui` and not be generic anymore. Other place that needs a size should use a `Vec2` instead has it's way more powerful.


So... my recommendations to document `bevy_math` is to remove all its modules, then it won't need that much doc 😄

mockersf reviewed (changes requested):
Is the `FaceToward` trait really needed? It seems it's only duplicate functionality from https://docs.rs/glam/latest/glam/f32/struct.Mat4.html#method.look_at_rh. It's also never used in Bevy code. There's also https://github.com/bitshifter/glam-rs/issues/214, @alice-i-cecile do you have more information?

Reading the doc on `Rect`, I find it very confusing. Looking through its usage in code:
* first thing I noticed, we actually have two different structs named `Rect`, the other one is defined in `bevy_sprite` and has a much more logical definition
https://github.com/bevyengine/bevy/blob/1d0d8a3397bd6fc2c14d42ffd0668d2443748912/crates/bevy_sprite/src/rect.rs#L8-L13
* the `Rect` struct in `bevy_math` is, as you noted, only used in `bevy_ui`, and it doesn't actually define a rectangle, only what's to do to each side of one (with of a border, width of a padding, ...)

I think the `Rect` from `bevy_math` should be moved to `bevy_ui` (and renamed), and it could probably lose the type parameter as I think it's only used with `Val`.

There is a pretty similar situation with the `Size` struct in `bevy_math`. It's only used as a generic `Vec2`, and that's only needed in `bevy_ui` as it's needed to have a `Vec2` over `Val`, which is not possible with the `Vec2` from `glam`. Most other places in code that need a size use directly a `Vec2`.

I think the `Size` struct should move to `bevy_ui` and not be generic anymore. Other place that needs a size should use a `Vec2` instead as it's way more powerful.


So... my recommendations to document `bevy_math` is to remove all its modules, then it won't need that much doc 😄

KDecay commented:
I agree with what you said.

I was also wondering what the point of the `FaceToward` Trait is, because it is never used and I was just able to remove the whole module without any errors. Maybe there are some people that use this functionality in their games, but I guess we have to go the same route as you did recently and just remove it and see if someone complains.

I'll start working on moving the geometry.rs to `bevy_ui` and renaming the `Rect.` Do you or anyone here have some recommendations on a new name that is more fitting and doesn't conflict with the `Rect` from `bevy_sprite`?

I will most likely remove the example which included `use bevy_math::prelude::*;` (see discussion above), since I don't want to add `use bevy::math::prelude::*;` with an ignore flag.

alice-i-cecile commented:
> Is the FaceToward trait really needed? It seems it's only duplicate functionality from https://docs.rs/glam/latest/glam/f32/struct.Mat4.html#method.look_at_rh. It's never used in Bevy code. There's also bitshifter/glam-rs#214, @alice-i-cecile do you have more information?

I agree! IMO we should make sure we've exposed that, and then use that functionality instead, deleting the `FaceForward` trait.

> I'll start working on moving the geometry.rs to bevy_ui and renaming the Rect.

I agree with this choice too. Perhaps `Padding` @KDecay?

> I will most likely remove the example which included use bevy_math::prelude::*; (see discussion above), since I don't want to add use bevy::math::prelude::*; with an ignore flag.

Agreed.

This all looks excellent: although the scope is larger, it's nice to fix up some of these code quality issues while we're here. Can you update the PR title @KDecay once you're done so it's easier to track in the git log? Something like "Documented and cleaned up bevy_math" would be fine.

MinerSebas commented:
The `bevy_sprite::Rect != bevy_math::Rect` fact was prevoisly brought up in #335.

KDecay commented:
> I agree with this choice too. Perhaps Padding?

That could work yeah. We came up with a simple UiRect in the discord, but Padding could also work. I'll probably just stick to UiRect for now. We can discuss the final name after I did all the points mentioned and the name seems inaccurate.

> The bevy_sprite::Rect != bevy_math::Rect fact was prevoisly brought up in #335.

Thanks for mentioning. The `Rect` of `bevy_math` will be moved to `bevy_ui` and also receive a rename to remove the duplicate type name. I'll consider and check the points mentioned in the issue and implement them as needed, so that we can close that issue.

> This all looks excellent: although the scope is larger, it's nice to fix up some of these code quality issues while we're here. Can you update the PR title once you're done so it's easier to track in the git log? Something like "Documented and cleaned up bevy_math" would be fine.

Yes of course. I'll update the title and also the objective so that it's clear what happened here.

KDecay commented:
# What I did

I removed every module inside of `bevy_math` so it is only a lib file which is using the `glam` crate.
I also removed the examples inside of the `bevy_math` lib file.
The code coverage of `bevy_math` is now at 100%.

The `FaceToward` trait got completely removed and the `Size` and `Rect` got moved to `bevy_ui`.
The `Rect` also received a rename and is now called `UiRect`.

@alice-i-cecile I decided to go with `UiRect` instead of `Padding` because the `UiRect` is used for positions, margins, paddings and borders, so calling it just `Padding` would be confusing.

# Removing `bevy_ui::Margins`

I checked #335 and the issue is still present. Inside of the `bevy_ui` crate there is a margins.rs which contains the `Margins` type. This type seems obsolete and should be replaced by the `UiRect` in my opinion, because the `UiRect` is also used for margins and could cause confusion. The only thing that `UiRect` isn't capable of is the `new(left: T, right: T, top: T, bottom: T)` function, which could easily be implemented, if needed.

Since we are removing a whole type we could also just ignore the `new()` function and keep it the way it currently is with defining it using `UiRect { .. }`. The `Margins` type is completely unused inside of bevy code so it is only used by users of bevy. If users use this type it would make it easier to migrate to `UiRect` if we added a `new()` function to `UiRect`. In my opinion we should do exactly this and add a `new()` function and remove the whole margins.rs file.

Let me know if we should do it and also if it should be inside of this PR or if we should create a new one!

alice-i-cecile commented:
I really like your proposed change; let's do this here so it doesn't get decoupled.

Thanks for the excellent work!

KDecay commented:
# What I did

I removed the margins.rs from `bevy_ui` and added the `new(left: T, right: T, top: T, bottom: T)`function to `UiRect`.

# Migration Guide

- old: `bevy_math::Rect` new: `bevy_ui::UiRect`
  - They are identical, but `bevy_ui::UiRect` has an additional `new(..)` function.
  - To migrate it is only required to change the path to the type and the type.
- old: `bevy_math::Size` new: `bevy_ui::Size`
  - They are identical.
  - To migrate it is only required to change the path to the type.
- old: `bevy_math::Mat4::face_toward(...)` new: `bevy_math::Mat4::look_at_rh(...)`
  - They should be identical (see https://docs.rs/glam/latest/glam/f32/struct.Mat4.html#method.look_at_rh).
  - To migrate it is only required to change the function name.
- old: `bevy_ui::Margins` new: `bevy_ui::UiRect`
  - They are not identical, but `bevy_ui::UiRect` has every functionality of `bevy_ui::Margins`.
  - To migrate it is only required to change the type name.
- The Trait `FaceToward` got removed.
  - This trait was only used for the `Mat4` type, but if any user used this in their own projects they have to create this trait themselves.
  - No migration possible.

james7132 commented:
>  No migration possible.

Probably should suggest to use https://docs.rs/glam/latest/glam/f32/struct.Mat4.html#method.look_at_rh instead, as stated above.

KDecay commented:
> Probably should suggest to use https://docs.rs/glam/latest/glam/f32/struct.Mat4.html#method.look_at_rh instead, as stated above.

True. This would only allow the users to use the `FaceToward` functionality of the `bevy_math::Mat4` type. So if someone has their own type which needs the `face_toward(...)`function they would still have to implement that themselves.

I edited my last comment so that we have at least some sort of migration guide to provide for users. Thanks for pointing that out!

alice-i-cecile commented on crates/bevy_math/src/lib.rs:
```diff
@@ -1,14 +1,16 @@
-mod geometry;
+#![warn(missing_docs)]
+
+//! Provides math types and functionalities for the bevy game engine.
```
```suggestion
//! Provides math types and functionality for the Bevy game engine.
```

alice-i-cecile commented on crates/bevy_math/src/lib.rs:
```diff
@@ -1,14 +1,16 @@
+
+//! Provides math types and functionalities for the bevy game engine.
+//!
+//! The commonly used types are vectors and matrices.
```
These should be linked :)

alice-i-cecile commented on crates/bevy_text/src/text.rs:
```diff
@@ -154,5 +154,5 @@ impl Default for TextStyle {
 #[reflect(Component)]
 pub struct Text2dSize {
-    pub size: Size,
+    pub size: Vec2,
```
I am unsure if I like the change away from `Size`. The more strongly-typed API is nice, and we can implement specialty methods on it later. Also named structs > tuple structs for clarity.

Other opinions?

alice-i-cecile replied:
Given that `Size` still exists, I would be in favor of changing this back.

alice-i-cecile replied:
Oh I see: this would pull in another dependency. Okay sure, that's totally fine with me then.

KDecay replied:
The following was said by @mockersf:

>There is a pretty similar situation with the Size struct in bevy_math. It's only used as a generic Vec2, and that's only needed in bevy_ui as it's needed to have a Vec2 over Val, which is not possible with the Vec2 from glam. Most other places in code that need a size use directly a Vec2.

> I think the Size struct should move to bevy_ui and not be generic anymore. Other place that needs a size should use a Vec2 instead as it's way more powerful.

And I agree to these suggestions, because otherwise we would have to add dependencies to `bevy_ui` everywhere we want to use the `Size` type. Because the `Size` type is literally just a generic and simpler version of a `Vec2` I went with using a `Vec2` instead. 

I kept every occurrence of `Size` inside of `bevy_ui` and only changed the ones outside of that crate. This is because we need a `Vec2` of type `Val` which is only possible with the `Size` struct. The other places don't need this functionality which is why I changed them to be `Vec2`s instead. 

One thing I just noticed is that he also mentioned that the struct shouldn't be generic anymore. I will change this as well.

alice-i-cecile commented on crates/bevy_ui/src/geometry.rs:
```diff
@@ -0,0 +1,361 @@
+///
+/// A margin is used to create space around UI elements, outside of any defined borders.
+///
+/// In this example an UI margin gets created. It has a left value of 10px, a top value of 20px,
```
```suggestion
/// In this example we are creating a UI margin. It has a left value of 10px, a top value of 20px,
```

It is hard to read text written in the passive voice in English, so I try to avoid it.

alice-i-cecile commented on crates/bevy_ui/src/geometry.rs:
```diff
@@ -0,0 +1,361 @@
+/// A position is used to determine where to place an UI element.
+///
+/// In this example an UI position gets created. It has a left value of 100px and a top value of 50px.
+/// If positioned absolutely this would correspond to an UI element which is positioned 100px to the right
```
What happens if both `left` and `right` are specified and don't match? This seems like a very strange use of this type, even if it doesn't appear to be introduced by you.

KDecay replied:
I tested this, because I was also curious about this. It seems that `top` and `left` have a higher priority than `bottom` and `right`. I could imagine that these values get checked and if they are not `Val::Undefined` they are used and whichever one gets checked first gets the priority, but that's just an idea. Haven't checked the source code of this.

KDecay replied:
I copied this question to my comment below so that we have a centralized point with all the open points.

alice-i-cecile reviewed (changes requested):
A few small nits :) I am mostly baffled by the use of `UiRect` for positioning; @mockersf do you have any insight here on why this pattern exists?

alice-i-cecile commented on crates/bevy_ui/src/geometry.rs:
```diff
@@ -0,0 +1,361 @@
+///
+/// # Example
+///
+/// In this example an UI size gets created. It has a width of 100px and a height of 200px.
```
```suggestion
```

I don't think this adds any value for the reader.

alice-i-cecile commented on crates/bevy_ui/src/geometry.rs:
```diff
@@ -0,0 +1,361 @@
+    ///
+    /// # Example
+    ///
+    /// In this example an UI size gets created. It has a width of 100px and a height of 200px.
```
```suggestion
```

alice-i-cecile commented on crates/bevy_ui/src/geometry.rs:
```diff
@@ -0,0 +1,361 @@
+///
+/// ## Position
+///
+/// A position is used to determine where to place an UI element.
```
```suggestion
/// A position is used to determine where to place a UI element.
```

English wants "a UI" and "a user interface". I know it starts with a vowel, but the "u" is hard, and uses a "y" sound. There's a few places where this happens if you search <3

KDecay replied:
Good to know. Thanks! I'll change these to `a`.

KDecay commented:
# Discussion Points

1. Should `Size<T>` and/or `UiRect<T>` be converted to `Size<Val>` and/or `UiRect<Val>`?
1. Why is a `UiRect` used for positions?
1. What happens when both `left` and `right` inside of the `UiRect` are specified and don't match?

# Current Situation

1. > I think the Size struct should move to bevy_ui and not be generic anymore.

@mockersf said that we should change `Size<T>` to `Size<Val>`. 

I'm currently not sure if we should, because using generics gives us great flexibility for the future. Currently `Size` is only used with `Val` values, so it would be possible to change it to `Size<Val>`.

2. > I am mostly baffled by the use of UiRect for positioning.

@alice-i-cecile asked why a `UiRect` is used for positioning. I don't know the UI code that well, but I think this is used because that way we can easily have different UI elements snap to different corners/sides of the window. 

For example we can define a position using a `UiRect` with a `left` value of `Val::Px(100.0)` and a `top` value of `Val::Px(50.0)` which would essentially anchor it to the top left corner of the screen. If we now define a position using a `UiRect` with a `right` value of `Val::Px(100.0)` and a `bottom` value of `Val::Px(50.0)` we anchor it to the bottom right corner of the screen.

I'm not sure how anchoring works with this, but if the position only uses two values depending on which `Anchor` is used then it might be worth considering to use a `Vec2` for a position/point instead.

3. > What happens if both left and right are specified and don't match?

@alice-i-cecile also asked what happens when the `left` and `right` value of a position / `UiRect` don't match. 

I tested this, because I was also curious about this. It seems that `top` and `left` have a higher priority than `bottom` and `right`. I could imagine that these values get checked in a particular order and if they are not `Val::Undefined` they are used and whichever one gets checked first gets the priority, but that's just an idea. Haven't checked the source code of this.

alice-i-cecile commented:
> 1.  I'm currently not sure if we should, because using generics gives us great flexibility for the future. Currently Size is only used with Val values, so it would be possible to change it to Size<Val>.

I think we should do this; `Val` can easily be extended with new units if needed, and users can write their own `From` impls for both `Size` and `Val`. Reducing generics where we don't need them is nice.

> 2. why is a UiRect is used for positioning. 

Alright, that explanation largely makes sense, although I'm still suspicious. Certainly not your problem here. Clarifying the docs here a bit further will be good enough.

> 3.  It seems that top and left have a higher priority than bottom and right

We should verify this, and then document it here :)

mockersf commented:
> 5\. > What happens if both left and right are specified and don't match?

This should be a totally normal case. When something is positioned with both right and left, its width should change.

KDecay commented:
> I think we should do this; Val can easily be extended with new units if needed, and users can write their own From impls for both Size and Val. Reducing generics where we don't need them is nice.

Alright. I'll change `Size<T>` to `Size<Val>` then. @alice-i-cecile I'm pretty sure that `UiRect<T>` is also only used as `UiRect<Val>`. Should I remove this generic as well if this is the case? 

> Alright, that explanation largely makes sense, although I'm still suspicious. Certainly not your problem here. Clarifying the docs here a bit further will be good enough.

I think I'll have to play around a little to see how the UI responds to different position settings. Hopefully I can come up with a decent explanation of the behaviour.

> We should verify this, and then document it here :)

Yup, I will combine this and the point above to verify this behaviour.

> This should be a totally normal case. When something is positioned with both right and left, its width should change.

Oh wow. I even knew this from other UI's I used. Not sure why I didn't think about this. Thanks for clarifying! I will also check if this is actually the case with our current UI.

alice-i-cecile commented:
> Alright. I'll change Size<T> to Size<Val> then. @alice-i-cecile I'm pretty sure that UiRect<T> is also only used as UiRect<Val>. Should I remove this generic as well if this is the case?

Yes please!

KDecay commented:
This PR is ready for review.

I checked the `bevy_math` and `bevy_ui` crate with the following commands, which ran through fine:
- `cargo clippy -p bevy_math --no-deps --all-targets --all-features -- -D warnings -A clippy::type_complexity -W clippy::doc_markdown`
- `cargo clippy -p bevy_ui --no-deps --all-targets --all-features -- -D warnings -A clippy::type_complexity -W clippy::doc_markdown`

While fixing the `clippy::doc_markdown` errors I copied the changes made in PR #3484.
So if this PR gets merged the PR #3484 can be closed (I added this to the PR comment as well).

# Migration Guide

- old: `bevy_math::Rect` new: `bevy_ui::UiRect`
  - They are identical, but `bevy_ui::UiRect` has an additional `new(..)` function.
  - To migrate it is only required to change the path to the type and the type.
- old: `bevy_math::Size` new: `bevy_ui::Size`
  - They are identical.
  - To migrate it is only required to change the path to the type.
- old: `bevy_math::Mat4::face_toward(...)` new: `bevy_math::Mat4::look_at_rh(...)`
  - They should be identical (see https://docs.rs/glam/latest/glam/f32/struct.Mat4.html#method.look_at_rh).
  - To migrate it is only required to change the function name.
- old: `bevy_ui::Margins` new: `bevy_ui::UiRect`
  - They are not identical, but `bevy_ui::UiRect` has every functionality of `bevy_ui::Margins`.
  - To migrate it is only required to change the type name.
- The Trait `FaceToward` got removed.
  - This trait was only used for the `Mat4` type, but if any user used this in their own projects they have to create this trait themselves.
  - To migrate we suggest using the `bevy_math::Mat4::look_at_rh` function or to create a trait that provides this functionality (see https://docs.rs/glam/latest/glam/f32/struct.Mat4.html#method.look_at_rh).
- old: `Size<T>` new: `Size<Val>`
  - Removed the generic trait so that `Size` can only be used with`Val`.
  - To migrate every instance of `Size` should now use `Val` which is the most common type to use. 
- old: `UiRect<T>` new: `UiRect<Val>`
  - Removed the generic trait so that `UiRect` can only be used with`Val`.
  - To migrate every instance of `UiRect` should now use `Val` which is the most common type to use.

mockersf reviewed (changes requested):
Looking good!

Most of the asserts in the `UiRect` doc code examples are asserting that rust is working and don't have anything to do with Bevy code (asserting on a struct creation, asserting on the result of a subtraction between floats). I don't think we should assert on those, if they ever fail anything Rust wouldn't be working anyway.

KDecay commented:
Very true.
I removed the really basic asserts and only left the ones in that test our implementations like `new()` and `all()`.
I also fixed the order of the `UiRect` fields in the examples so that it is consistent throughout the docs.

alice-i-cecile reviewed (approved):
I've done a final pass on this and it looks great! 

@cart, this is blending code quality improvements / reorganization with docs to a larger degree than I'm comfortable merging myself under the standard arrangement. It's also probably helpful for you to have an updated mental model of this area of the code base.

Could you merge this yourself when you get a chance?

KDecay commented:
Going to close this down and split it into multiple smaller PR's to make it easier to review and merge.
